### PR TITLE
IGNITE-13713 [ML]: Add target encoding preprocessor

### PR DESCRIPTION
**Pull request by mrk-andreev**
* Add target encoding preprocessor

Issue: https://issues.apache.org/jira/browse/IGNITE-13713

### CODE REVIEW

File: examples/src/main/java/org/apache/ignite/examples/ml/preprocessing/encoding/TargetEncoderExample.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+                    .withEncodedFeatures(targetEncodedfeaturesIndexies)
+                    .minSamplesLeaf(1)
+                    .minCategorySize(1L)
+                    .smoothing(1d)
+                    .fit(ignite,
+                        dataCache,
+                        strEncoderPreprocessor
+                    );
+
+                Preprocessor<Integer, Object[]> lbEncoderPreprocessor = new EncoderTrainer<Integer, Object[]>()
```
Comment (zaleslaw) [inline examples/src/main/java/org/apache/ignite/examples/ml/preprocessing/encoding/TargetEncoderExample.java:98]: Sorry, but I didn't understand this pipeline? Why those 3 encoders are combined here? Could they work only in this combination?
In my opinion, user have a choice what to do with Strings, but he should choose one method (not the chain of methods).

Please share your vision here

File: modules/ml/src/test/java/org/apache/ignite/ml/preprocessing/encoding/TargetEncoderPreprocessorTest.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            new HashSet() {
+                {
+                    add(0);
+                    add(1);
+                    add(2);
+                }
+            });
+
+        double[][] postProcessedData = new double[][] {
+            {1.0, 0.1, 1.0},
```
Comment (zaleslaw) [inline modules/ml/src/test/java/org/apache/ignite/ml/preprocessing/encoding/TargetEncoderPreprocessorTest.java:82]: Could you explain please numbers in the last columns: why are they 1.0 and 2.0? not 0.33 and 0.66

File: modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/encoding/target/TargetEncoderPreprocessor.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+ *
+ * The Target Encoder Preprocessor encodes string values (categories) to double values
+ * in range [0.0, 1], where the value will be presented as a regularized mean target value of
+ * a category.
+ *
+ * alpha = 1 / (1 + exp(-(categorySize - min_samples_leaf) / beta))
+ * encodedValue = globalTargetMean * (1 - alpha) + categoryTargetMean * alpha
+ * if categorySize == 1 then use globalTargetMean
+ *
+ * min_samples_leaf - minimum samples to take category average into account.
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/encoding/target/TargetEncoderPreprocessor.java:37]: looks like min_samples_leaf is not used in this class

File: modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/encoding/EncoderTrainer.java [hunk 4]
```diff
@@ -116,6 +143,77 @@
...
+            }
+        );
+
+        TargetEncodingMeta[] targetEncodingMetas = new TargetEncodingMeta[targetCounters.length];
+        for (int i = 0; i < targetCounters.length; i++) {
+            if (handledIndices.contains(i)) {
+                int finalI = i;
+
+                targetEncodingMetas[i] = new TargetEncodingMeta(
+                    targetCounters[i].getTargetSum() / targetCounters[i].getTargetCount(),
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/encoding/EncoderTrainer.java:187]: I suggest to refactor constructor parameters to separate variables for readability purposes

File: modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/encoding/EncoderTrainer.java [hunk 5]
```diff
@@ -116,6 +143,77 @@
...
+                    targetCounters[i].getCategoryTargetSum().entrySet().stream()
+                        .collect(Collectors.toMap(
+                            Map.Entry::getKey,
+                            value -> {
+                                double prior = targetCounters[finalI].getTargetSum() /
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/encoding/EncoderTrainer.java:192]: Also this lambda should be encapsulated and commented separately

File: modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/encoding/EncoderTrainer.java [hunk 6]
```diff
@@ -116,6 +143,77 @@
...
+                                    double smoove = 1 / (1 +
+                                        Math.exp(-(categorySize - minSamplesLeaf) / smoothing));
+                                    return prior * (1 - smoove) + categoryMean * smoove;
+                                }
+                            }
+                        ))
+                );
+            }
+        }
+
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/encoding/EncoderTrainer.java:213]: remove the blank line

File: modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/encoding/EncoderTrainer.java [hunk 7]
```diff
@@ -343,6 +441,81 @@ else if (lbVal instanceof Double)
...
+                String strVal;
+                Object featureVal = row.features().getRaw(i);
+
+                if (featureVal.equals(Double.NaN)) {
+                    strVal = EncoderPreprocessor.KEY_FOR_NULL_VALUES;
+                    row.features().setRaw(i, strVal);
+                }
+                else if (featureVal instanceof String)
+                    strVal = (String)featureVal;
+                else if (featureVal instanceof Double)
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/encoding/EncoderTrainer.java:472]: Maybe add type conversion to Doulbe from another Number types (and boolean)

**Review by zaleslaw — CHANGES_REQUESTED** [review]
So, this PR makes me happy, but raises a few questions:
1) Use-case with GBT looks strange for me
2) So large dataset in 32 000 rows looks big for resources too. Could we use another dataset instead of proposed? Titanic or something else with 100-1000 rows. 

Let's discuss it here, in this PR

File: examples/src/main/java/org/apache/ignite/examples/ml/preprocessing/encoding/TargetEncoderExample.java (hunk 1 above)
Comment (mrk-andreev) [reply examples/src/main/java/org/apache/ignite/examples/ml/preprocessing/encoding/TargetEncoderExample.java:98]: I think that we want use `EncoderType.TARGET_ENCODER` only for a few columns (may be only one). In this example I use `EncoderType.STRING_ENCODER` as general propose encoder and `EncoderType.TARGET_ENCODER` for special one.

File: modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/encoding/target/TargetEncoderPreprocessor.java (hunk 3 above)
Comment (mrk-andreev) [reply modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/encoding/target/TargetEncoderPreprocessor.java:37]: Yes, but it is used for evaluate `TargetEncodingMeta`, so I wanted to mention this in encoder class.

File: modules/ml/src/test/java/org/apache/ignite/ml/preprocessing/encoding/TargetEncoderPreprocessorTest.java (hunk 2 above)
Comment (mrk-andreev) [reply modules/ml/src/test/java/org/apache/ignite/ml/preprocessing/encoding/TargetEncoderPreprocessorTest.java]: Described each case

File: modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/encoding/EncoderTrainer.java (hunk 4 above)
Comment (mrk-andreev) [reply modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/encoding/EncoderTrainer.java]: Extracted to new method

File: modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/encoding/EncoderTrainer.java (hunk 5 above)
Comment (mrk-andreev) [reply modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/encoding/EncoderTrainer.java]: `prior` evaluation extracted but lambda still exists.

File: modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/encoding/EncoderTrainer.java (hunk 6 above)
Comment (mrk-andreev) [reply modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/encoding/EncoderTrainer.java]: Done

File: modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/encoding/EncoderTrainer.java (hunk 7 above)
Comment (mrk-andreev) [reply modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/encoding/EncoderTrainer.java]: Add `Number` & `Boolean`

**Pull request merged**